### PR TITLE
tests: document the test suites in the snapcraft repo

### PR DESCRIPTION
**Pull request by elopio**

### CODE REVIEW

File: TESTING.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+## Manual tests
+
+Snapcraft has a few manual tests documented in [manual-tests.md](manual-tests.md).
+These are tests for features that are not possible to automate, or that the complexity of automating them is too high while the time it takes to run them by hand is not too much.
+
+We try very hard to automate as much as possible and to keep the manual test suite as small as possible. Please do not add a manual test before discussing it with the team, and make sure that there are very good reasons not to automate it.
+
+### Staging server
+
+Snapcraft has the ability to upload snaps for publication in the Snappy Store. If you're working on a feature that requires you to interact with the store, you might want to use the staging server instead of the production store. To do
```
Comment (kalikiana) [inline TESTING.md]: You're not wrapping the lines anymore... if that's intentional, this should be discussed and applied separately.

File: TESTING.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+To run all the tests execute:
+
+    ./runtests.sh
+
+You can selectively run only one of the suites, and apply a filter:
+
+    ./runtests.sh [static|unit|integration|plugins|store|snaps] [pattern]
+
+Examples:
```
Comment (kalikiana) [inline TESTING.md]: I'm missing examples for static, integration and snaps.

File: TESTING.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+### Running the tests
+
+To run all the tests execute:
+
+    ./runtests.sh
+
+You can selectively run only one of the suites, and apply a filter:
+
+    ./runtests.sh [static|unit|integration|plugins|store|snaps] [pattern]
```
Comment (kalikiana) [inline TESTING.md]: Instead of this here, it'd be nicer to have that as a Usage message in the script, and be able to say "check `./runtests.sh --help` for the exact syntax", also because you wouldn't need to search for this documentation later.

File: TESTING.md (hunk 1 above)
Comment (elopio) [inline TESTING.md]: This was the only paragraph with wrapping like that in HACKING.md. I was just trying to make it consistent here. I'm -1 on moving it to a separate PR.

File: TESTING.md (hunk 3 above)
Comment (elopio) [inline TESTING.md]: https://bugs.launchpad.net/snapcraft/+bug/1702716

File: TESTING.md (hunk 2 above)
Comment (elopio) [inline TESTING.md]: I added a few more lines.

File: TESTING.md (hunk 1 above)
Comment (kalikiana) [inline TESTING.md]: You're right. In that case I'm fine with it.

File: TESTING.md (hunk 3 above)
Comment (kalikiana) [inline TESTING.md]: Cheers

File: TESTING.md [hunk 4]
```diff
@@ -110,3 +110,50 @@ The integration and snaps suites can be run using the snapcraft source from the
 The store tests by default will start fake servers that are configured to reply like the real store does. But you can run them also against the staging and production store servers. To do that, you will need to set the `TEST_STORE` environment variable to either `staging` or `production`, and you also have to pass credentials for a valid user in that store with the environment variable `TEST_USER_EMAIL` and `TEST_USER_PASSWORD`, like this:
 
     TEST_STORE=staging TEST_USER_EMAIL=test@example.com TEST_USER_PASSWORD=Hola123* ./runtests.sh store [pattern]
+
+## Autopkgtests for the snapcraft deb
+
+Autopkgtests are tests for the project packaged as a deb. The unit tests are run during autopkgtests while the snapcraft deb is being built. Then the resulting deb is installed, and the integration and snaps suites are executed using the installed snapcraft.
+
```
Comment (kalikiana) [inline TESTING.md]: Lovely! Thanks for adding those as well.

File: TESTING.md (hunk 1 above)
Comment (sergiusens) [inline TESTING.md]: but why it `To do` on its own separate line?

File: TESTING.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  * To run only the full unit test suite:
+
+    ```
+    ./runtests.sh unit
+    ```
+
+  * To run only the unit tests in the file called `test_init.py`:
+
+    ```
+    ./runtests.sh unit test_init.py
```
Comment (sergiusens) [inline TESTING.md]: missing ```

Comment (sergiusens) [inline TESTING.md]: but you don't need ``` if you are going to indent (comment applies for all these invokations).

File: TESTING.md (hunk 1 above)
Comment (elopio) [inline TESTING.md]: fixed.

File: TESTING.md (hunk 5 above)
Comment (elopio) [inline TESTING.md]: wow, thanks, I didn't see it. I've added them.
You need the quotes if you want to keep them inside the bullet block.

Comment (jamesbeedy) [inline TESTING.md]: @elopio can integration tests  be ran individually in this way also?

Comment (elopio) [inline TESTING.md]: @jamesbeedy individual files, yes. To run an individual test, you will have to use python3 -m unittest directly.

**Comment by jamesbeedy** [comment]
@elopio reading through this, it's answering a lot of the questions I was having when initially getting my feet wet running tests. Thanks for the cycles here!
